File: 930.cpp
struct Solution {
    int numsBelow(int ub, vector<int> &nums){
        if (ub < 0){
            return 0;
        }
        int res = 0, subsum = 0, e = 0;
        for (int s = 0; s < nums.size(); ++s){
            while (e < nums.size() && subsum + nums[e] <= ub){
                subsum += nums[e++];
            }
            res += e - s;
            subsum -= nums[s];
        }
        return res;
    }
    inline int numSubarraysWithSum(vector<int>& nums, int goal) {
        return this->numsBelow(goal, nums) - this->numsBelow(goal - 1, nums);
    }
};

File: sol/930/930.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        res = 0
        prefixSum = 0
        sumCount = {0: 1}
        for n in nums:
            prefixSum += n
            res += sumCount.get(prefixSum - goal, 0)
            sumCount[prefixSum] = sumCount.get(prefixSum, 0) + 1
        return res
